File: frontend/react_web/src/utils/supportAccess.ts
import type { User } from '@/store/auth';

export const ASSIGNEE_UPDATE_STATUSES = ['IN_PROGRESS', 'RESOLVED', 'CLOSED'] as const;

export function assigneeStatusOptions<T extends { value: string }>(all: T[]): T[] {
  return all.filter((s) => (ASSIGNEE_UPDATE_STATUSES as readonly string[]).includes(s.value));
}

export function defaultAssigneeStatus(current: string): string {
  return (ASSIGNEE_UPDATE_STATUSES as readonly string[]).includes(current)
    ? current
    : 'IN_PROGRESS';
}

export function isPlatformSuperAdmin(user: User | null | undefined): boolean {
  if (!user || user.role !== 'SUPER_ADMIN') return false;
  const tenantId = user.tenantId || user.tenant?.id || 'default';
  return tenantId === 'default';
}

export function isDefaultTenantSupportAssignee(user: User | null | undefined): boolean {
  if (!user || user.isClient) return false;
  if (user.role === 'SUPER_ADMIN' || user.role === 'ADMIN') return false;
  const tenantId = user.tenantId || user.tenant?.id || 'default';
  const slug = user.tenant?.slug;
  return tenantId === 'default' && (!slug || slug === 'default');
}

export function canAccessSupportIssues(user: User | null | undefined): boolean {
  return isPlatformSuperAdmin(user) || isDefaultTenantSupportAssignee(user);
}

export function canReportProblem(user: User | null | undefined): boolean {
  if (!user || user.isClient) return false;
  return !isPlatformSuperAdmin(user);
}

export function assigneeCanUpdateIssueStatus(options: {
  isSuperAdmin: boolean;
  ticketStatus: string;
  assigneeIds: string[];
  userId?: string;
}): boolean {
  if (options.isSuperAdmin) return true;
  if (options.ticketStatus === 'CLOSED') return false;
  if (!options.userId) return false;
  return options.assigneeIds.includes(options.userId);
}

File: frontend/flutter_apps/shared_core/lib/src/support_access.dart
/// Support inbox / Issues access helpers (mirrors web `supportAccess.ts`).
const assigneeUpdateStatuses = ['IN_PROGRESS', 'RESOLVED', 'CLOSED'];

List<Map<String, dynamic>> assigneeStatusOptions(
  List<Map<String, dynamic>> all,
) {
  return all
      .where((s) => assigneeUpdateStatuses.contains(s['value']?.toString()))
      .toList();
}

String defaultAssigneeStatus(String current) {
  return assigneeUpdateStatuses.contains(current) ? current : 'IN_PROGRESS';
}

bool assigneeCanUpdateIssueStatus({
  required bool isSuperAdmin,
  required String ticketStatus,
  required List<String> assigneeIds,
  required String? userId,
}) {
  if (isSuperAdmin) return true;
  if (ticketStatus == 'CLOSED') return false;
  if (userId == null) return false;
  return assigneeIds.contains(userId);
}
